portal/trigger_states/adherence_cache_validation.py

```diff
@@ -0,0 +1,162 @@
+            match = re.findall(r'\d+', row.rs_id_visit)
+            if not match:
+                raise ValueError(f"Patient {self.patient_id} has bogus rs_id_visit value: {r2.rs_id_visit}")
+            this_patient_months[int(match[-1])] = row.status
+            if row.status == 'Withdrawn':
+                self.withdrawal_month = int(match[-1])
+
+        self.adherence_data = this_patient_months
+        if self.withdrawal_month:
+            return
+
+        # if withdrawal_month wasn't found, calculate now if user has withdrawn
+        patient = unchecked_get_user(self.patient_id, allow_deleted=True)
+        _, withdrawal = consent_withdrawal_dates(patient, EMPRO_RS_ID)
+        if not withdrawal:
+            return
+        query = text(
+            "select at from qb_timeline where qb_id = 28 and status = 'due' "
+            "and research_study_id = 1 and user_id = :user_id")
+        result = db.engine.execute(query, {"user_id": self.patient_id}).first()[0]
+        withdrawal_month = -1
+        while True:
+            if withdrawal < result:
+                self.withdrawal_month = withdrawal_month
+                return
+            result += timedelta(days=30)
+            withdrawal_month += 1
+            assert withdrawal_month > 12
+
+    def trigger_states_months_by_patient(self):
+        this_patient_ts_months = {}
+        query = text(
+            "select visit_month, state from trigger_states where user_id = :patient_id order by id"
+        )
+        for row in db.engine.execute(query, {"patient_id":self.patient_id}):
+            # allow to overwrite improved state given order
+            visit_month = row.visit_month + 1  # 0 index, align with adherence
+            this_patient_ts_months[visit_month] = state_map[row.state]
+        self.ts_data = this_patient_ts_months
+
+    def report(self, message):
+        if not self.report_called:
+            self.report_called = True
+            print(f"Differences for {self.patient_id} adherence | trigger states")
+        print(message)
+
+    def show_differences(self):
+        if self.patient_id == 9090:
+            pass
+            #import pdb; pdb.set_trace()
+        if self.adherence_data is None and self.ts_data is None:
+            return None
+        for i in range(1,13):
+            ad = self.adherence_data.get(i)
+            ts = self.ts_data.get(i)
+            if ad is None and ts is None:
+                continue
+            if ad and not ts:
+                if ad in ('Due', 'Expired', 'Overdue'):
+                    # trigger states often never started w/o user intervention
+                    continue
+                if ad == 'Withdrawn':
+                    # trigger states often not generated after withdrawal
+                    continue
+                if self.patient_id == 5903:
+                    # one of a kind, missing trigger_states rows for visit 11
```

Comment:
instead of referencing this issue by patient ID, would it be possible to ensure there's no more than 1 of these?